src/flamegpu/simulation/CUDAEnsemble.cu

```diff
@@ -199,11 +211,274 @@ unsigned int CUDAEnsemble::simulate(const RunPlanVector &plans) {
+                                fprintf(stderr, "\rWarning: Run %u failed on rank %d, device %d, thread %u with exception: \n%s\n",
+                                    e_detail.run_id, world_rank, e_detail.device_id, e_detail.runner_id, e_detail.exception_string);
+                            }
+                        }
+                        run_id = detail::MPISimRunner::Signal::RequestJob;
+                    }
+                    if (run_id == detail::MPISimRunner::Signal::RequestJob) {
+                        r.store(next_run++);
+                    }
+                }
+                // Check whether MPI runners require a job assignment
+                MPI_Iprobe(
+                    MPI_ANY_SOURCE,           // int source
+                    EnvelopeTag::RequestJob,  // int tag
+                    MPI_COMM_WORLD,           // MPI_Comm communicator
+                    &flag,                    // int flag
+                    &status);                 // MPI_Status*
+                while (flag) {
+                    // Receive the message (kind of redundant as we already have the status and it carrys no data)
+                    memset(&status, 0, sizeof(MPI_Status));
+                    MPI_Recv(
+                        nullptr,                  // void* data
+                        0,                        // int count
+                        MPI_CHAR,                 // MPI_Datatype datatype (can't use MPI_DATATYPE_NULL)
+                        MPI_ANY_SOURCE,           // int source
+                        EnvelopeTag::RequestJob,  // int tag
+                        MPI_COMM_WORLD,           // MPI_Comm communicator
+                        &status);                 // MPI_Status*
+                    // Respond to the sender with a job assignment
+                    MPI_Send(
+                        &next_run,               // void* data
+                        1,                       // int count
+                        MPI_UNSIGNED,            // MPI_Datatype datatype
+                        status.MPI_SOURCE,       // int destination
+                        EnvelopeTag::AssignJob,  // int tag
+                        MPI_COMM_WORLD);         // MPI_Comm communicator
+
+                    if (next_run >= plans.size()) ++mpi_runners_fin;
+                    ++next_run;
+                    // Check again
+                    MPI_Iprobe(MPI_ANY_SOURCE, EnvelopeTag::RequestJob, MPI_COMM_WORLD, &flag, &status);
+                }
+                // Print progress to console
+                if (config.verbosity >= Verbosity::Default && next_run_reported != next_run) {
+                    const int progress = static_cast<int>(next_run) - static_cast<int>(TOTAL_RUNNERS * world_size);
+                    fprintf(stdout, "\rCUDAEnsemble progress: %d/%u", progress < 0 ? 0 : progress, static_cast<unsigned int>(plans.size()));
```

Comment:
I dislike the usage of carriage returns it breaks if a model includes any stdout output. 

They make logging the output to file store messy info, and if a printf is included in the model at all it breaks output (admitidily with concurrent ensembles this will be out of order anyway)

Not going to fight its inclusion, just logging my dislike.

Reply:
Find me the relevant code to detect the stream type to allow output file to be mutated and i'll make the change. Iirc last time I tried, it didn't work.

Reply:
Actually given the use-case for MPI (distribtued batch jobs) then just getting rid of the carraige returns makse more sense than having them at all. 

I.e. it's always going to be batch jobs on hpc systems, where logging to disk will happen, so no carraige returns is the better option. 

---

should be `isatty` from `unistd.h` iirc. it's part of the posix spec though not c/c++, so would need guarding appropriately, but depending on how programs are executed it might not report a differnt value than expected (I think).

Reply:
fair point.